org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/ExecutorExtension.java

```diff
@@ -0,0 +1,107 @@
+package org.osgi.test.junit5;
+
+import static org.junit.platform.commons.util.AnnotationUtils.findAnnotatedFields;
+import static org.junit.platform.commons.util.ReflectionUtils.isPrivate;
+import static org.junit.platform.commons.util.ReflectionUtils.makeAccessible;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Parameter;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionConfigurationException;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.junit.platform.commons.util.ExceptionUtils;
+import org.junit.platform.commons.util.ReflectionUtils;
+
+public class ExecutorExtension implements AfterEachCallback, BeforeEachCallback, ParameterResolver {
```

Comment:
Although it perhaps goes beyond the scope of osgi-test, there is a lot about proper osgi testing that involves concurrency. I think we should consider making these executor-related extensions/rules part of the public API. There are also some concurrency-related utilities I wrote as part of the bnd tester testing that have generic application (eg, waiting for another thread to enter the wait state) that could form part of a concurrency testing library. That can be a topic for another issue/PR though; happy for it to be test-only scope for now.

Reply:
issue created.